roles/download/tasks/download_file.yml

```diff
@@ -4,43 +4,121 @@
     msg:
       - "URL: {{ download.url }}"
       - "Dest: {{ download.dest }}"
+  run_once: true
 
-- name: file_download | Create dest directory
+- set_fact:
+    cache_dir: "{{ local_release_dir }}/{{ download.dest | regex_replace('^\\/', '') }}"
+  tags:
+    - facts
+
+- name: file_download | Create local cache directory
   file:
-    path: "{{download.dest|dirname}}"
+    path: "{{ cache_dir | dirname }}"
+    state: directory
+    recurse: yes
+  delegate_to: localhost
```

Comment:
There's two types of directories being made in this task. One on the nodes and one on the local machine. The local machine folder does not depend on the delegate, because even if a delegate downloads files / images, they'll be downloaded to the local machine, for caching, if download_run_once is true. On a side note: I was contemplating adding an 'always_cache_local' flag to toggle this feature, so that downloading to local cache can be turned off. But I had already made all preparations for the pull request by then and didn't want to 'boil the ocean', ie. I wanted to get the working and tested code out before doing any more modifications to the code. (please don't say: please add that toggle first... please ;-)

Reply:
downloading locally should only be done when, delegate_download == localhost, 
I don't see any usecase we would download locally, then sync to download_delegate, that would then again sync to all other nodes.

Reply:
>I don't see any usecase we would download locally, then sync to download_delegate, that would then again sync to all other nodes.

That is correct, that situation will not occur. What will occur, is this:

- starting situation: local cache is empty, ie. no files/images in cache
- download_run_once=true, download_local=false
- vagrant up --> nodes boot up
- delegate node downloads images / files
- files/images get copied back to localhost for later caching
- files/images get pushed from delegate node to other nodes
- kubespray does its thing...
- vagrant destroy (because... laptop OS upgrade, dual boot other OS, RAM shortage etc...)
- vagrant up --> nodes boot up
- locally cached files/images get pushed to delegate node
- delegate node downloads/pulls updated images/files, if needed
- (updated) files/images get copied back to localhost for later caching
- files/images get pushed from delegate node to other nodes
- etc.

This is the main feature of this pull request: No local docker instance is needed to cache container images and (non container image) files also get cached.

Reply:
this seems very specific to the usecase, everything is local (vagrant). 
Vagrant is more a 'test'/'demo' thing and not a end, which is deploy a production setup. 
As such roles should not be more complex to accomodate vagrant deployment. 

We are actually trying to reduce the number of options of deployment rather than increasing (e.g local cache). 

I think 'download_local', should be removed, and only keep 'delegate_download'. 

If you delegate_download to localhost, what is not working for your usecase ?

Reply:
I haven't reviewed the delegate_download at all. so maybe it's broken, not complete as it should be. 

But I see two modes: 
1. Each nodes download the dependencies (can be an issue for very large cluster and add dependencies to external/3rd party services (like gcr.io/dockerhub/github)....
2. Download all dependencies once in one of node(download_delagate), and synchronize files to all other nodes
I still doesn't scale very well, but improve dependencies. 

I the case of vagrant/download_local, the download_delegate would be localhost.

Reply:
I agree that there should be no excessive complexity. Once there is a clear idea on how to progress, we / I can start putting some more time in this code again, don't you agree? Maybe you can use the new PR, along with your current opinion on how to progress, to determine what bits of code to keep and how to develop this code, going forward. Let me know if and how I can help: https://github.com/kubernetes-incubator/kubespray/pull/3720

On the other hand I find kubespray to be a very useful tool to have on a laptop, for development and educational purposes, so having a relatively fast provisioning system is important in that aspect. The caching of the hyperkube binary alone saves me a 240MB download on each kubespray deployment, which amounts to five minutes of download time on some links I'm connected to. Maybe I can fix this issue by moving the local caching bit to vagrant. That shouldn't be too hard and it will remove the complexity of that from kubespray main code and put it in support code, ie. vagrant.

Reply:
It still not clear, why is configuring download_delegate: localhost  not working in your case ?

Reply:
Well, it is not a question if it is not working. The old code works fine. The new code just works differently. It copies back files / images to localhost, after they have been downloaded on the delegate node. For this, the cache directory must be available on the localhost. So in this case, the cache directory is not dependent on delegate_host; it is always localhost.

On each new run, the new code checks if a file is in the local cache, and if so, it copies it to the delegate node. That way the delegate node does not need to re-download all the docker images and / or files, each time kubespray is started. This is only useful for use on laptops / with vagrant.